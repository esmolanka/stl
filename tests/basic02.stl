type Product = { fst : Unit, snd : Void }
#eval Product

type Sum = < Inl : Unit | Inr : Void >
#eval Sum

type SumOfProducts =
  | Left  : Product
  | Right : Product

#eval SumOfProducts

type ProductOfSumsOptional =
  { first?  : Sum
  , second? : Sum
  }
#eval ProductOfSumsOptional

type ProductOfSums =
  { first  : Sum
  , second : Sum
  }
#eval ProductOfSums

type ProductOfFunctions =
  { choice  : Unit -> Void -> Sum
  , product : Unit -> Void -> Product
  }
#eval ProductOfFunctions
