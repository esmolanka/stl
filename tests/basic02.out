tests/basic02.types:2:1-2:14:
  Normalised:
    ∃ (ρ : Row). {fst : Unit, snd : Void | ρ}
  Kind:
    Type
tests/basic02.types:5:1-5:10:
  Normalised:
    ∀ (ρ : Row). <Inl? : Unit | Inr? : Void | ... ρ>
  Kind:
    Type
tests/basic02.types:11:1-11:20:
  Normalised:
    ∀ (ρ : Row).
      < Left? : ∃ (ρ : Row). {fst : Unit, snd : Void | ρ}
      | Right? : ∃ (ρ : Row). {fst : Unit, snd : Void | ρ}
      | ... ρ
      >
  Kind:
    Type
tests/basic02.types:17:1-17:28:
  Normalised:
    ∃ (ρ : Row).
      { first? : ∀ (ρ : Row). <Inl? : Unit | Inr? : Void | ... ρ>
      , second? : ∀ (ρ : Row). <Inl? : Unit | Inr? : Void | ... ρ>
      | ρ
      }
  Kind:
    Type
tests/basic02.types:23:1-23:20:
  Normalised:
    ∃ (ρ : Row).
      { first : ∀ (ρ : Row). <Inl? : Unit | Inr? : Void | ... ρ>
      , second : ∀ (ρ : Row). <Inl? : Unit | Inr? : Void | ... ρ>
      | ρ
      }
  Kind:
    Type
tests/basic02.types:29:1-29:25:
  Normalised:
    ∃ (ρ : Row).
      { choice :
          Unit -> Void -> ∀ (ρ : Row). <Inl? : Unit | Inr? : Void | ... ρ>
      , product : Unit -> Void -> ∃ (ρ : Row). {fst : Unit, snd : Void | ρ}
      | ρ
      }
  Kind:
    Type